### ISSUE TITLE
Unable to inject objects into initial active routes using `instanceInitializer`

### ISSUE THREAD

**Issue by bmac** [emberjs/ember.js#11172]
When attempting to remove deprecation warnings in Ember Data. I moved the initializer code that [creates the store](https://github.com/emberjs/data/blob/master/packages/ember-data/lib/initializers/store.js) into an `instanceInitializer`.

Unfortunately, it seems the routes that are initially active (e.g. `ApplicationRoute` and `IndexRoute` if you visit `/`) when the page is loaded are created before any `instanceInitializer`s get a change to run. 

It looks like the ordering of steps in the `didBecomeReady` function causes the problem. The [startRouting call](
https://github.com/emberjs/ember.js/blob/257c2bc8dcc4c6f418abf6b1142c688d9dcee0c7/packages/ember-application/lib/system/application.js#L727) ends up creating Route instances, while the `instanceInitializer` are not run until after the [bootResolver is resolved](https://github.com/emberjs/ember.js/blob/257c2bc8dcc4c6f418abf6b1142c688d9dcee0c7/packages/ember-application/lib/system/application.js#L738) later in the function.

#### Example JS Bin
http://emberjs.jsbin.com/yexehoperu/1/edit?js,console

**Comment by amk221** [comment]
_think_ I've come across the same issue. I was going to file a bug, but I'm not 100% sure.
@bmac would you mind awfully clarifying if this is the same issue: http://jsfiddle.net/amk221/yxg92rvL/

**Comment by bmac** [comment]
Looks like the same issue @amk221

**Comment by amk221** [comment]
Ok thank you!

**Comment by jamesarosen** [comment]
I believe I'm running into something similar.

On a (non-instance) initializer, `application.inject` exists, but emits a deprecation warning:
```js
// app/initializers/foo.js
export default {
  name: 'foo',
  initialize: function(registry, application) {
    application.inject('controller', 'foo', 'service:foo'); // works, but warns
  }
};
```
Here, `application` looks like
```js
{
  $: function() {},
  LOG_ACTIVE_GENERATION: false,
  LOG_VIEW_LOOKUPS: false,
  Router: my-app.Router,
  __container__: Container,
  ... // dozens more properties
}
```
But on an _instance_ initializer, `application.inject` doesn't exist at all:
```js
// app/instance-initializers/foo.js
export default {
  name: 'foo',
  initialize: function(application) {
    application.inject('controller', 'foo', 'service:foo'); // throws!
  }
};
```
Here, `application` looks like
```js
{
  __ember1431740202170: null,
  __ember_meta__: Object,
  __nextSuper: undefined,
  _didSetupRouter: true,
  applicationRegistry: Registry,
  container: Container,
  customEvents: null,
  registry: Registry,
  rootElement: "#ember-testing"
}
```
Those two `application` objects are very different.

**Comment by jamesarosen** [comment]
Ah. My problem is because the second isn't an instance of [`Application`](https://github.com/emberjs/ember.js/blob/v1.12.0/packages/ember-application/lib/system/application.js); it's an instance of [`ApplicationInstance`](https://github.com/emberjs/ember.js/blob/v1.12.0/packages/ember-application/lib/system/application-instance.js).

That's... confusing.

**Comment by amk221** [comment]
I believe there are a few issues here:

1. [11172](https://github.com/emberjs/ember.js/issues/11172#issuecomment-102570176) The original issue whereby you're unable to inject into routes
2. The instance is not an application.<br>deferReadiness [isn't available](http://jsfiddle.net/amk221/jvc05Ly2/) (fiddle), but according to [this blog post](http://emberjs.com/blog/2015/05/13/ember-1-12-released.html#toc_instance-initializers) it should be.
3. [11114](https://github.com/emberjs/ember.js/issues/11114) There is no documentation on these changes

**Comment by btecu** [comment]
It would be nice to get this fixed in **1.12.1**.

**Comment by rwjblue** [comment]
@btecu - I would happily ship a 1.12.x with a fix if a PR were submitted.  Have time to take a look?
